db/write_batch_test.cc

```diff
@@ -899,6 +904,89 @@ TEST_F(WriteBatchTest, MemoryLimitTest) {
   ASSERT_TRUE(s.IsMemoryLimit());
 }
 
+TEST_F(WriteBatchTest, AssignTimestamps) {
+  // We assume the last eight bytes of each key is reserved for timestamps.
+  // Therefore, we must make sure each key is longer than eight bytes.
+  constexpr size_t key_size = 16;
+  constexpr size_t num_of_keys = 10;
+  std::vector<std::string> key_strs;
+  for (size_t i = 0; i < num_of_keys; ++i) {
+    key_strs.emplace_back(std::string(key_size, '\0'));
+  }
+  ColumnFamilyHandleImplDummy cf1(0);
+  ColumnFamilyHandleImplDummy cf2(4, test::ComparatorWithU64Ts());
+  ColumnFamilyHandleImplDummy cf3(5, test::ComparatorWithU64Ts());
+
+  WriteBatch batch;
+  // Write to the batch. We will assign timestamps later.
+  for (const auto& key_str : key_strs) {
+    ASSERT_OK(batch.Put(&cf1, key_str, "value"));
+    ASSERT_OK(batch.Put(&cf2, key_str, "value"));
+    ASSERT_OK(batch.Put(&cf3, key_str, "value"));
+  }
+
+  const auto checker1 = [](uint32_t cf, size_t& ts_sz) {
+    if (cf == 4 || cf == 5) {
+      ts_sz = sizeof(uint64_t);
+    } else if (cf == 0) {
+      ts_sz = 0;
+    } else {
+      return Status::Corruption("Invalid cf");
+    }
+    return Status::OK();
+  };
+  ASSERT_OK(
+      batch.AssignTimestamp(std::string(sizeof(uint64_t), '\0'), checker1));
```

Comment:
We could consider using a timestamp that is not all zero bytes here and checking whether it was actually set (same below)

Reply:
That may require a new `WriteBatchHandler` subclass, is this something you have in mind?

Reply:
Yeah, I suppose there's no easier way, right? For testing purposes, it might be sufficient to override `PutCF` though?

Reply:
For this test yes. Let me add it.

Reply:
Added.